mautic/mautic
# TPROD-432 : Install chosen.js via npm

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [cleanup]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes https://mautic.atlassian.net/browse/TPROD-432 



#### Description:



We need to install chosen.js via npm. There were 2 hacks which were preventing us from install this package via npm. 
1. https://github.com/mautic/mautic/blob/986f05790776669b4fcb1720dbcac5762ad7ec19/app/bundles/CoreBundle/Assets/js/libraries/3a.chosen.jquery.js#L398-L428
2. https://github.com/mautic/mautic/blob/986f05790776669b4fcb1720dbcac5762ad7ec19/app/bundles/CoreBundle/Assets/js/libraries/3a.chosen.jquery.js#L495-L509

This PR introduces `patch-package` which was installed by `npm install patch-package --save-dev` command so it is available for dev environments only. With patch-package we can create many patches for packages which have custom code like above without forking the actual package repo and waiting endlessly for the patch to be applied in the actual repo.

#### To use `patch-package`:
1. Edit the actual `node_modules/<package>/<file>` files with your desired patch.
2. Run `npx patch-package <package>`, replace `<package>` with your package, in this case `npx patch-package chosen-js`. This will generate a patch file in the `patches` folder. Commit this file for others to use.
3. To apply the patch in your code run `npx patch-package`.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Check select boxes through out the application whether they work as before. I found Segment filters page the best to look for such select boxes.

#### Quickest dev testing:
1. On your dev environment, delete `node_modules`.
2. Run composer install (it will execute `post-install-cmd` scripts which has `npm ci` and `npx patch-package`)
3. Ensure that `node_modules/chosen-js/chosen.jquery.js` has the patched code at the two places as mentioned above.
